### ISSUE TITLE
torch.unique(t, dim=1, return_counts=True) does not work

### ISSUE THREAD

**Issue by aifartist** [pytorch/pytorch#148743]
### 🐛 Describe the bug

torch.unique(t, dim=1, return_counts=True) does not work
The "value" result still has dup's for each row(dim=1) and only one set of counts is returned instead of one for each row.

```
import torch

t = torch.tensor([[1, 1, 7, 3, 7, 7, 3], [2, 2, 2, 6, 6, 4, 3]])

print(torch.unique(t, dim=1, return_counts=True))
for tRow in t: # Show what the results should be
    print(torch.unique(tRow, return_counts=True))

print(torch.unique_consecutive(t, dim=1, return_counts=True))
for tRow in t:  # Show what the results should be
    print(torch.unique_consecutive(tRow, return_counts=True))
```

The repo also has the results if unique or unique_consecutive is called once per row with a for loop.  As can be seen the results are different.

I can't make any sense of the results.  Basically worthless for 2d.  However, it is clear having unique() work across multiple rows using dim=1 isn't straight forward because the number of unique values might differ from row to row and the resultant values or counts wouldn't be rectangular.

HOWEVER, one could 'pad' the results to return:
uniqvals = [[1, 3, 7, <pad>], [2, 3, 4, 6]]
counts = [[2, 2, 3, 0], [3, 1, 1, 2]]
<pad> could be any value of the dtype, and where count=0 is how we'd know a position was just padding.
IMO, it is an important use case to go through a large number of 1d vectors and compute unique on each without a for loop on a gpu.
Thinking ahead, to know how much to pad, one would need to know the maximum number of unique values on any given row.  To save needing to do extra work to handle padding after the first pass one could just set a max for the length of the output "rows".  To guarantee nothing is missed just specify the len of that dimension from the input.  Hopefully I made this part clear.

NOTE:  I disagree with: https://github.com/pytorch/pytorch/issues/130217
I find the simple 1d case understandable and usable. But with an additional dimension(s) it isn't that it couldn't be understandable but it simply does not work.

### Versions

Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Feb  4 2025, 14:48:35) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-53-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 570.86.15
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.7.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-3,11,13-31
Off-line CPU(s) list:                 4-10,12
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i9-13900K
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   21
Socket(s):                            1
Stepping:                             1
CPU(s) scaling MHz:                   56%
CPU max MHz:                          5800.0000
CPU min MHz:                          0.0000
BogoMIPS:                             5990.40
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            752 KiB (21 instances)
L1i cache:                            1.2 MiB (21 instances)
L2 cache:                             26 MiB (9 instances)
L3 cache:                             36 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-3,11,13-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] came-pytorch==0.1.3
[pip3] mypy-extensions==1.0.0
[pip3] nexfort==0.1.dev329+torch251cu124
[pip3] numpy==2.1.3
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] oldest-supported-numpy==2023.12.21
[pip3] pytorch-fid==0.3.0
[pip3] torch==2.6.0
[pip3] torchaudio==2.5.1
[pip3] torchprofile==0.0.4
[pip3] torchvision==0.21.0
[pip3] triton==3.2.0
[conda] Could not collect

**Comment by mattduerrmeier** [comment]
If you don't return the counts, how do you know that the value used as `pad` is not present in the tensor? What you want to do is apply the function along an axis, like you do in the for loop.
Numpy `unique` has the same behavior as PyTorch:
```
import torch
import numpy as np

a = torch.tensor([[1, 1, 7, 3, 7, 7, 3], [2, 2, 2, 6, 6, 4, 3]])
print(torch.unique(a, dim=1, return_counts=True))
print(np.unique(a.numpy(), axis=1, return_counts=True))
```

Output:
```
(tensor([[1, 3, 3, 7, 7, 7],
        [2, 3, 6, 2, 4, 6]]), tensor([2, 1, 1, 1, 1, 1]))
(array([[1, 3, 3, 7, 7, 7],
       [2, 3, 6, 2, 4, 6]]), array([2, 1, 1, 1, 1, 1]))
```

**Comment by aifartist** [comment]
> If you don't return the counts, how do you know that the value used as `pad` is not present in the tensor? What you want to do is apply the function along an axis, like you do in the for loop. Numpy `unique` has the same behavior as PyTorch:
> 
> ```
> import torch
> import numpy as np
> 
> a = torch.tensor([[1, 1, 7, 3, 7, 7, 3], [2, 2, 2, 6, 6, 4, 3]])
> print(torch.unique(a, dim=1, return_counts=True))
> print(np.unique(a.numpy(), axis=1, return_counts=True))
> ```
> 
> Output:
> 
> ```
> (tensor([[1, 3, 3, 7, 7, 7],
>         [2, 3, 6, 2, 4, 6]]), tensor([2, 1, 1, 1, 1, 1]))
> (array([[1, 3, 3, 7, 7, 7],
>        [2, 3, 6, 2, 4, 6]]), array([2, 1, 1, 1, 1, 1]))
> ```

Just because numpy returns the wrong result is no reason for torch.unique() to.
Those returned values are not unique.
Given the dimension returning the unique values is not guaranteed to have rows all with the same length(without padding) it becomes a must to set return_counts to True to know for sure.  But I'd rather do that than get duplicates returned from a function which is "documented" as returning "UNIQUE" values.
And there is a second way to get this to work without needing return_counts=True.  If you are guaranteed the values don't include the pad value you can also get this to work.  Consider a set of values from 1 to 52 for a deck of cards.  You could pad with 0, -1, 255 or whatever.
A third way would be to pad with the option of returning the len/dimension of each row in a tensor.  Thus, an alternative to return_counts=True like "return_nunique=True".

I want a unique() which works as documented even if it requires a "pad=<val>" argument.

**Comment by albanD** [comment]
Hey!

I think the main confusion here is that torch.unique definition in PyTorch is not what you would like it to be.
So it is "correct" from our point of view but "incorrect" for you.

I don't expect we will change the semantic of the current torch.unique() to do what you're asking for for backward compatibility reasons.

We could have a new op that does what you want, but, as you pointed out, it is a tricky up to create as each row might have a different number of elements.

**Comment by aifartist** [comment]
> torch.unique definition in PyTorch is not what you would like it to be.
No.  I want the *implementation* to match *YOUR* *definition* provided in
the doc's.
Unique means "unique".  In the test case I provided they aren't unique and
the counts are worthless in that I have no idea what they mean in the 2d
case I've shown.

*Backwards compatibility for what appears to be a junk result is hardly a
worthy goal.*

However, if a meaningful definition of some kind of *useful* result for the
>1 dimension case were provided in the documentation and the results
matched that definition then OK.  I've seen this kind of thing happen with
complex SQL queries where I'm an old time database internals architect.
Trust me, the semantics can really get ugly and correctness is a must in
databases.

The question is do you even see that there is a problem between the
definition and the result for the > 1 dimension case?  Then with clarity of
thought, on a complex to grasp thing, can we come up with a solution.  Yes,
"torch.reallyunique2d()" is an option but it feels silly UNLESS there is
usefulness in with the current ?mess? but if so then I want to hear THAT
definition.

Thanks,
Dan, MSFT, Amazon, Salesforce, Oracle, Sybase, IBM, ...

**Comment by aifartist** [comment]
To be clear, the possibility exists that I'm misunderstanding something in
the documentation that makes the actual behavior clear, predictable, and
useful.  My comments are based on that not being the case.  I can be
convinced with an explanation of the semantics that I am missing.
